pilot/pkg/proxy/envoy/v2/ads.go

```diff
@@ -354,13 +354,13 @@ func (s *DiscoveryServer) removeCon(conID string, con *XdsConnection) {
 }
 
 func (s *DiscoveryServer) pushRoute(con *XdsConnection) error {
-	rc := [][]*xdsapi.RouteConfiguration{}
+	rc := []*xdsapi.RouteConfiguration{}
+	// TODO: once per config update
+	services, err := s.env.Services()
 	for _, rn := range con.Routes {
-		r, err := s.ConfigGenerator.BuildRoutes(s.env, *con.modelNode, rn)
-		if err != nil {
-			log.Warnf("ADS: config failure, closing grpc %v", err)
-			return err
-		}
```

Comment:
can you restore this method and stick the buildSidecarOutbounbdHttpRouteConfig in the implementation of BuildRoutes ? I can add logic for handling ingress/gateway there. This way, the code here remains unchanged..

Reply:
The problem is I'm not sure a coarse interface is best for what we do. But I guess less of a problem for routes, listeners and clusters are the main issue.

Reply:
Sure, but we can refine the interface over time and always consume from interface. It feels a bit clunky to make existing functions public and break through the interface barrier. When we have v1alpha4/v1beta1, this will be another big headache to transition all over the code, as opposed to just changing a single point in the interface.

Reply:
+1 I'd prefer to stick it into the existing interface. In fact, I didn't see you changed the field in discovery.go from the interface type to the concrete implementation. Please do not do that. Extend the interface, don't depend on some concrete type. The whole goal is to make this more modular and pluggable.

Reply:
Once again - designing a good interface is tricky, and at this point we don't have enough data or time (or priority) to define such interface. 

Once we have time to agree on an interface - that takes into account the event-based goals and 
the proper caching - we can change back. 

I replaced the interface with concrete type because I think the interface needs to be removed until we have a good, agreed one.

Reply:
In other words: in 0.9 we'll finish the optimizations started in this PR (of having List called only on config change - instead of many times deep inside each generator ), and we'll have a better idea of what dataset is needed to generate each config. This will allow us to also know what configs to re-generate and push for each config event. 

When we have that - and caching of the 'mesh-wide' configs ( outbound cluster rules, outbound clusters ) and caching for the per-deployment configs ( in listeners and clusters ) - we can define a proper interface, with the right granularity.  

And as a side effect: it will be easier to test the generators - since they'll get their data from the caller instead of calling env.List.